XenAdmin/Dialogs/ControlDomainMemoryDialog.cs

```diff
@@ -0,0 +1,177 @@
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR 
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, 
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING 
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE 
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF 
+ * SUCH DAMAGE.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Drawing;
+using System.Text;
+using System.Windows.Forms;
+using XenAdmin.Actions;
+using XenAdmin.Commands;
+using XenAPI;
+
+namespace XenAdmin.Dialogs
+{
+    public partial class ControlDomainMemoryDialog : XenDialogBase
+    {
+        private Host host;
+        private Host_metrics hostMetrics;
+        private double origMemory;
+
+        public const int MAXIMUM_DOM0_MEMORY_GB =  256;
+
+        public ControlDomainMemoryDialog(Host host)
+            : base(host.Connection)
+        {
+            if (host == null) throw new ArgumentNullException("host");
+
+            InitializeComponent();
+            this.host = host;
+            this.host.PropertyChanged += Server_PropertyChanged;
+            hostMetrics = connection.Resolve(this.host.metrics);
+            if (hostMetrics != null)
+                hostMetrics.PropertyChanged += Server_PropertyChanged;
+            Text = string.Format(Messages.CONTROL_DOMAIN_MEMORY_DIALOG_TITLE, this.host.Name);
+            Populate();
+        }
+
+        private void UpdateMaintenanceWarning()
+        {
+            Host_metrics metrics = host.Connection.Resolve(host.metrics);
+            bool maintenanceMode = host.MaintenanceMode || (metrics != null && !metrics.live);
+
+            maintenanceWarningImage.Visible = maintenanceWarningLabel.Visible = maintenanceModeLinkLabel.Visible = !maintenanceMode;
+            hostRebootWarningImage.Visible = hostRebootWarningLabel.Visible = maintenanceMode;
+
+            memorySpinner.Enabled = maintenanceMode;
+        }
+
+        private void Populate()
+        {
+            VM vm = host.ControlDomainZero;
+
+            // Since updates come in dribs and drabs, avoid error if new max and min arrive
+            // out of sync and maximum < minimum.
+            if (vm.memory_dynamic_max >= vm.memory_dynamic_min &&
+                vm.memory_static_max >= vm.memory_static_min)
+            {
+                double min = vm.memory_static_min;
+                double max = Math.Min(host.memory_available_calc, MAXIMUM_DOM0_MEMORY_GB * Util.BINARY_GIGA);
+                double value = vm.memory_dynamic_min;
+                // Avoid setting the range to exclude the current value: CA-40041
+                if (value > max)
+                    max = value;
+                if (value < min)
+                    min = value;
+                memorySpinner.Initialize(Messages.CONTROL_DOMAIN_MEMORY_LABEL, null, value, max);

```

Comment:
As discussed, the memory spinner needs to be refactored to ensure that whenever we initialise it or change the range we don't find ourselves with a value out of range due to rounding issues. Agreed that this will be fixed in a different PR.